jgrapht-core/src/main/java/org/jgrapht/graph/AsSynchronizedGraph.java

```diff
@@ -0,0 +1,1283 @@
+/*
+ * (C) Copyright 2018-2018, by CHEN Kui and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.graph;
+
+import java.io.*;
+import java.lang.*;
+import java.util.*;
+import java.util.concurrent.locks.*;
+import java.util.function.*;
+import java.util.stream.*;
+import org.jgrapht.*;
+
+/**
+ * Create a synchronized (thread-safe) Graph backed by the specified Graph. This Graph is designed
+ * to read and write mutually exclusive. In order to guarantee serial access, it is critical that
```

Comment:
"designed to support concurrent reads which are mutually exclusive with writes.'